Improve wallet fee logic and fix GUI bugs

morcos opened the pull request:
This builds on #10589  (first 5 commits from that PR, last 5 commits are new)

The first couple commits refactor to use the CCoinControl class to pass fee calculation parameters around. 

This allows for fixing the buggy interaction in QT between the global payTxFee which can be modified by the RPC call settxfee or temporarily modified by the QT custom fee settings.  Before these changes the GUI could sometimes send a transaction with a recently set payTxFee and not respect the settings displayed in the GUI.   After these changes, using the GUI does not involve the global transaction confirm target or payTxFee.

The prospective fee displays in the smart fee slider and the coin control dialog are changed to use the fee calculation from GetMinimumFee, this simplifies the code and makes them slightly more correct in edge cases.

Maxing the fee calculation with the mempool min fee is move from estimateSmartFee to GetMinimumFee.

This fixes a long standing bug, and should be tagged for 0.15 as it is holding up finalizing the estimatesmartfee RPC API before release.

morcos committed:
Make CoinControl a required argument to CreateTransaction

morcos committed:
Refactor to use CoinControl in GetMinimumFee and FeeBumper

Improve parameter precedence in coin_control

morcos committed:
Use CoinControl to pass custom fee setting from QT.

This fixes buggy behavior where we were temporarily setting and unsetting the
global payTxFee when trying to send a transaction with a custom fee from the
GUI. The previous behavior was inconsistent depending on the order of using the
RPC call settxfee and clicking various radio buttons in the sendcoinsdialog.
The new behavior is that transactions sent with the GUI will always use either
the smartfee slider value or the custom fee set on the GUI and they will not
affect the global defaults which are only for RPC and initial GUI values.

morcos committed:
Make QT fee displays use GetMinimumFee instead of estimateSmartFee

Remove helper function (CalculateEstimateType) for determining whether
estimates should be conservative or not, now that this is only called
once from GetMinimumFee and incorporate the logic directly there.

morcos committed:
Remove checking of mempool min fee from estimateSmartFee.

This check has been moved to the wallet logic GetMinimumFee. The rpc call to
estimatesmartfee will now no longer return a result maxed with the mempool min
fee, but automated fee calculations from the wallet will produce the same result
as before and coincontrol and sendcoins dialogs in the GUI will correctly
display the right prospective fee.

changes to policy/fees.cpp include a big whitespace indentation change.

TheBlueMatt reviewed (commented):
Looks good, definitely want this to go in with #10589 (I'm super not a fan of CalculateEstimateType as an GetMinimumFee helper, but its easy to get rid of after this PR, less so after 10589).

morcos commented on src/qt/walletmodel.cpp:
```diff
@@ -667,8 +669,9 @@ bool WalletModel::bumpFee(uint256 hash)
+        CCoinControl no_coin_control;
         LOCK2(cs_main, wallet->cs_wallet);
-        feeBump.reset(new CFeeBumper(wallet, hash, nTxConfirmTarget, false, 0, true));
+        feeBump.reset(new CFeeBumper(wallet, hash, no_coin_control, 0));
```
Good catch.  Will fix. I think I just assumed how the code worked and didn't even look at what I was replacing.

morcos commented on src/wallet/coincontrol.h:
```diff
@@ -19,12 +21,12 @@ class CCoinControl
+    //! Override the default payTxFee if set, 0 = use smart fee estimation
+    boost::optional<CFeeRate> m_feerate;
+    //! Override the default confirmation target
+    boost::optional<unsigned int> m_confirm_target;
```
I hate the magic values of 0.  Actually I think I should make the logic even slightly more clear.
There should just be a clear parameter precedence.
- coin_control.m_feerate
- coin_control.m_confirm_target
- ::payTxFee (global)
- ::nTxConfirmTarget (globa)

It goes down the list until it fines one that is set and all the first 3 should be boost::optional's.  Why do you not like boost::optional?
I could skip changing the global payTxFee for now, and still have that work via the 0 magic value.
The code to process this will be slightly longer in `GetMininumFee` but I think it'll be far more clear no?

TheBlueMatt replied:
Hmm, maybe I shouldnt argue against boost::optional. I'm not a fan of optional types, but they do make sense here. Any further simplification would be nice, this stuff is gross (thanks for cleaning it up!).

morcos commented on src/wallet/coincontrol.h:
```diff
@@ -18,14 +21,16 @@ class CCoinControl
-    //! Override the default confirmation target, 0 = use default
-    int nConfirmTarget;
+    //! Override the default payTxFee if set, 0 = use smart fee estimation
+    boost::optional<CFeeRate> m_feerate;
```
(see above, will document when we have a final logic)

morcos commented:
Rebased against the updated #10589 and fixed bug as well as improved parameter precedence logic for Coin Control.

ryanofsky commented on src/wallet/coincontrol.h:
```diff
@@ -41,9 +43,9 @@ class CCoinControl
         fAllowWatchOnly = false;
         setSelected.clear();
-        nFeeRate = CFeeRate(0);
+        m_feerate = boost::none;
```
In commit "Refactor to use CoinControl in GetMinimumFee and FeeBumper"

Maybe replace ` = boost::none` with `.reset()` here and other places to remove boost reference and make it a little easier to port to `std::optional` in the future.

morcos replied:
will do

morcos replied:
Actually turns out `.reset()` is deprecated so left it as is.

ryanofsky commented on src/wallet/coincontrol.h:
```diff
@@ -19,12 +21,12 @@ class CCoinControl
     bool fAllowWatchOnly;
-    //! Override estimated feerate
+    //! Override automatic min/max checks on fee, m_feerate must be set if true
     bool fOverrideFeeRate;
```
In commit "Refactor to use CoinControl in GetMinimumFee and FeeBumper"

You should definitely get rid of the `fOverrideFeeRate` member now that `m_feerate` is optional.

morcos replied:
Well..  almost
`fOverrideFeeRate` is used from `fundrawtransaction` and it seems to me that it might make sense that you don't want your specified fee rate subject to clamping by various min fees or the maxTxFee when you are using it from there but you do generally.  

In any case it would be a change of functionality to change that, so we should do it separately.

morcos commented on src/wallet/wallet.cpp:
```diff
@@ -2567,7 +2567,7 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
         {
             std::vector<COutput> vAvailableCoins;
-            AvailableCoins(vAvailableCoins, true, coinControl);
+            AvailableCoins(vAvailableCoins, true, &coin_control);
```
I agree, but decided it would make this PR a bit too big.

Which multiply defined values are you referring to though?

morcos commented:
Rebased cleanly so only new commits are left after #10589 was merged

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -2719,17 +2718,11 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
-                CAmount nFeeNeeded = GetMinimumFee(nBytes, currentConfirmationTarget, ::mempool, ::feeEstimator, &feeCalc, false /* ignoreGlobalPayTxFee */, conservative_estimate);
-                if (coinControl && coinControl->fOverrideFeeRate)
-                    nFeeNeeded = coinControl->nFeeRate.GetFee(nBytes);
+                // Allow to override automatic fee calculation (min/max checks) over coin control instance
```
It would be nice to move this check into GetMinimumFee so that everything is ompletely in the same place.

morcos replied:
done

morcos commented:
Rebased to accomodate #10712 and moved fOverrideFeeRate inside GetMinimumFee

ryanofsky commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -642,16 +642,14 @@ void SendCoinsDialog::updateSmartFeeLabel()
-                                                                std::max(CWallet::fallbackFee.GetFeePerK(), CWallet::GetRequiredFee(1000))) + "/kB");
+    ui->labelSmartFee->setText(BitcoinUnits::formatWithUnit(model->getOptionsModel()->getDisplayUnit(), feeRate.GetFeePerK()) + "/kB");
+
+    if (feeCalc.reason == FeeReason::FALLBACK) {
```
In commit "Make QT fee displays use GetMinimumFee instead of estimateSmartFee":

Is this sufficient to know fee estimation succeeded? For example would it be possible for estimate to fail, and then fallbackFee to be less than RequiredFee, so reason would be REQUIRED not FALLBACK?

Also can you expand commit description to say a little bit about why it's better to display minimum fee instead of estimated fee directly here. It seems perfectly reasonable, I'd just like to understand why the min fee wasn't being being used all along, and whether one value is clearly more useful than the other or if this is more intended to be a minor tweak that simplifies the code.

morcos replied:
I think the behavior in rare edge cases is maybe not precisely correct, but there is likely no harm in it.  The only thing you are missing is a warning that fee estimation is not up to date yet, and before this PR that could already happen if there was a mempool min fee (in which case you wouldn't even get maxed with the fallback fee).  After this PR, you only miss the warning if your fallback fee is less than mempool min fee or minRelayTxFee or minTxFee.  The first case is a strict improvement over prior behavior. The second two cases can only result from very poor configuration of command line options.  In short, I think it could be improved to report even more precise information to the user but it's outside the scope of what I wanted to do here.

I can expand commit description, but the idea is the displays should correspond to the fees that your actually about to put on the transaction if you click send.  I think it's essentially a bug that it didn't do that before.

ryanofsky replied:
> I think the behavior in rare edge cases is maybe not precisely correct, but there is likely no harm in it.

I see. This is less significant than I thought because the fee is shown either way, just the labeling is different.

> I can expand commit description, but the idea is the displays should correspond to the fees that your actually about to put on the transaction if you click send. I think it's essentially a bug that it didn't do that before.

Thanks, I didn't know if the motivation was to simplify code or if min fee was actually preferable to display. Fact that min fee is used to send the transaction definitely makes sense as a reason to display it.

ryanofsky reviewed (commented):
utACK fb98c64134a3334d4d892af5ff08bc4291ea7b82

MIGUELWAXX commented:
cheers

TheBlueMatt commented:
utACK fb98c64134a3334d4d892af5ff08bc4291ea7b82

morcos commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2924,14 +2918,11 @@ UniValue bumpfee(const JSONRPCRequest& request)
-            newConfirmTarget = options["confTarget"].get_int();
-            if (newConfirmTarget <= 0) { // upper-bound will be checked by estimatefee/smartfee
+            int target = options["confTarget"].get_int();
+            if (target <= 0) { // FIXME: Check upper bound too
```
Correct, that comment about upper-bound is not true, I don't know why it was put there.
At the time I wrote these PR's I had a lot of important stuff I wanted to make sure got in, so I didn't want to fix every last little thing, but I think I could easily add a commit now that checks the bounds everywhere necessary.

morcos commented on src/qt/coincontroldialog.cpp:
```diff
@@ -587,7 +587,7 @@ void CoinControlDialog::updateLabels(WalletModel *model, QDialog* dialog)
         dFeeVary = (double)std::max(CWallet::GetRequiredFee(1000), payTxFee.GetFeePerK()) / 1000;
     else {
-        dFeeVary = (double)std::max(CWallet::GetRequiredFee(1000), ::feeEstimator.estimateSmartFee(coinControl->nConfirmTarget, NULL, ::mempool, conservative_estimate).GetFeePerK()) / 1000;
+        dFeeVary = (double)std::max(CWallet::GetRequiredFee(1000), ::feeEstimator.estimateSmartFee(*coinControl->m_confirm_target, NULL, ::mempool, conservative_estimate).GetFeePerK()) / 1000;
```
seems reasonable

morcos replied:
this one went away, but addressed a similar one in qt/sendcoinsdialog.cpp

morcos committed:
Properly bound check conf_target in wallet RPC calls

morcos commented:
Addressed @ryanofsky feedback and added a commit to do proper bounds checking.  boost::none's were just all removed in one commit for simplicity.

([feelogic.ver2](https://github.com/morcos/bitcoin/commits/feelogic.ver2)) -> c494da9 ([feelogic.ver2.squash](https://github.com/morcos/bitcoin/commits/feelogic.ver2.squash))

ryanofsky reviewed (commented):
utACK c494da90aa73051f3a21630f40ea84220e6e8bbb. Changes were adding qt coincontrol asserts, removing uses of boost::none, and adding a new commit with more conf_target checking.

jnewbery commented on src/rpc/mining.h:
```diff
@@ -12,4 +12,7 @@
 /** Generate blocks (mine) */
 UniValue generateBlocks(std::shared_ptr<CReserveScript> coinbaseScript, int nGenerate, uint64_t nMaxTries, bool keepScript);
 
+/** Check bounds on a command line confirm target */
```
Why is this in `rpc/mining.cpp`? All of the calls to this function are in `wallet/rpcwallet.cpp`. Can you just put the function there?

Also the comment is a bit misleading - this function checks confirm target in RPCs, not command line arguments.

morcos replied:
#10707 adds it to the RPC calls in rpc/mining.cpp and rpcwallet.cpp already includes rpc/mining.h

jnewbery reviewed (commented):
utACK c494da90aa73051f3a21630f40ea84220e6e8bbb . I'm not very familiar with the qt code, so I can't give you much feedback there, but the changes all look sensible.

A few nits. All could be addressed after merge (or left entirely).

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -2936,33 +2925,61 @@ CAmount CWallet::GetRequiredFee(unsigned int nTxBytes)
         if (feeCalc) feeCalc->reason = FeeReason::PAYTXFEE;
+        // Allow to override automatic min/max check over coin control instance
+        if (coin_control.fOverrideFeeRate) return fee_needed;
+    } else {
```
I wondered same thing initially, but I think it was written this way to unify the logic behind cases #2 and #4 which are very similar.

morcos commented:
@jnewbery See if you like the reorganization of GetMinimumFee and if so I'll squash both commits

jnewbery commented:
> See if you like the reorganization of GetMinimumFee

I wasn't necessarily advocating for changing the order of the conditionals, just for flattening the nested ifs. I'm happy in either order.

Definitely like the extra comments referring to which of the 4 branches the blocks match up to.

The indentation on LL2942-2943 is incorrect. Fix that and I'm happy!

morcos commented:
addressed @jnewbery nits and slight refactor of logic in `GetMinimumFee` (same end result)
([feelogic.ver4](https://github.com/morcos/bitcoin/commits/feelogic.ver4)) -> 759db91 ([feelogic.ver4.squash](https://github.com/morcos/bitcoin/commits/feelogic.ver4.squash))

morcos commented:
Unfortunately this required a bit of a rebase due to conflict with #10769 in `qt/sendcoinsdialog.cpp`
Other files had clean rebase.

gmaxwell reviewed (commented):
concept ACK

sipa commented on src/rpc/mining.cpp:
```diff
@@ -815,7 +815,6 @@ UniValue estimatesmartfee(const JSONRPCRequest& request)
             "\n"
             "A negative value is returned if not enough transactions and blocks\n"
             "have been observed to make an estimate for any number of blocks.\n"
-            "However it will not return a value below the mempool reject fee.\n"
```
Perhaps worth pointing out that the result may not be sufficient to be acceptable for the mempool?

morcos replied:
I think we should at least flag the change in the release notes but I don't think that's necessarily something users need to worry about.  In the event that the tx isn't accepted to the mempool it's still likely'ish to be confirmed within the target by being resubmitted after mempool min fee decays.  Which makes me wonder if we should even be worrying about mempool min fee in the GetMinimumFee logic but enough changes for now.

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -2932,33 +2922,52 @@ CAmount CWallet::GetRequiredFee(unsigned int nTxBytes)
+        // Allow to override automatic min/max check over coin control instance
+        if (coin_control.fOverrideFeeRate) return fee_needed;
+    }
+    else if (!coin_control.m_confirm_target && ::payTxFee != CFeeRate(0)) { // 3. TODO: remove magic value of 0 for global payTxFee
```
Nit: `else` on the same line

sipa reviewed (commented):
Concept ACK. I very much like making everything uniformly use a coin control object for guiding estimation, and making the GUI's coin control independent from global settings.

Code looks good, but I haven't had time to review all logic changes.

TheBlueMatt commented:
re-utACK 11590d39b9888403ead8354302e308eca139ba17

gmaxwell reviewed (approved):
ACK

laanwj commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -274,14 +268,10 @@ void SendCoinsDialog::on_sendButton_clicked()
     CCoinControl ctrl;
     if (model->getOptionsModel()->getCoinControlFeatures())
         ctrl = *CoinControlDialog::coinControl;
```
Thanks. Definitely seems an improvement with regard to CCoinControl handling.
Hopefully in a next PR we can finally get rid of this global object too.

laanwj commented:
utACK 11590d3
